File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/io/ByteBufferInputStream.java
package com.sedmelluq.discord.lavaplayer.tools.io;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;

/**
 * A byte buffer exposed as an input stream.
 */
public class ByteBufferInputStream extends InputStream {
  private final ByteBuffer buffer;

  /**
   * @param buffer The buffer to read from.
   */
  public ByteBufferInputStream(ByteBuffer buffer) {
    this.buffer = buffer;
  }

  @Override
  public int read() throws IOException {
    if (buffer.hasRemaining()) {
      return buffer.get() & 0xFF;
    } else {
      return -1;
    }
  }

  @Override
  public int read(byte[] array, int offset, int length) throws IOException {
    if (buffer.hasRemaining()) {
      int chunk = Math.min(buffer.remaining(), length);
      buffer.get(array, offset, length);
      return chunk;
    } else {
      return -1;
    }
  }

  @Override
  public int available() throws IOException {
    return buffer.remaining();
  }
}

File: main/src/main/java/lavaplayer/tools/io/ByteBufferInputStream.kt
package lavaplayer.tools.io

import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import kotlin.experimental.and

/**
 * A byte buffer exposed as an input stream.
 *
 * @param buffer The buffer to read from.
 */
class ByteBufferInputStream(private val buffer: ByteBuffer) : InputStream() {
    @Throws(IOException::class)
    override fun read(): Int {
        return if (buffer.hasRemaining()) {
            (buffer.get() and 0xFF.toByte()).toInt()
        } else {
            -1
        }
    }

    @Throws(IOException::class)
    override fun read(array: ByteArray, offset: Int, length: Int) = if (buffer.hasRemaining()) {
        val chunk = buffer.remaining().coerceAtMost(length)
        buffer[array, offset, length]
        chunk
    } else {
        -1
    }

    @Throws(IOException::class)
    override fun available(): Int =
        buffer.remaining()
}
